pkg/tlsutil/tls.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsutil
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CertType defines the type of the cert.
+type CertType int
+
+const (
+	// ClientCert defines a client cert.
+	ClientCert CertType = iota
+	// ServingCert defines a serving cert.
+	ServingCert
+	// ClientAndServingCert defines both client and serving cert.
+	ClientAndServingCert
+)
+
+// CertConfig configures how to generate the Cert.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Serving, client or both; defaults to both.
+	CertType CertType
+	// Optional CommonName is the common name of the cert; defaults to "".
+	CommonName string
+	// Optional Organization is Organization of the cert; defaults to "".
+	Organization []string
+	// Optional CA Key, if user wants to provide custom CA key via a file path.
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA cert via file path.
+	CACert string
+}
+
+// CertGenerator is an operator specific TLS tool that generates TLS assets for the deploying application.
+type CertGenerator interface {
+	// GenerateCert generates a secret containing TLS encryption key and cert, a Secret
+	// containing the CA key, and a ConfigMap containing the CA Certificate given the Custom
+	// Resource(CR) "cr", the Kubernetes Service "Service", and the CertConfig "config".
+	//
+	// The passed in "service" represents endpoint(s) of a communicating party. We assume that the
+	// deployed applications are typically communicated via the Kubernete Service. Therefore, the
+	// cert is generated to certify the endpoint(s) defined by the "service".
```

Comment:
```Go
	// The passed in "service" represents endpoint(s) of a communicating party. We assume that the
	// deployed applications are typically communicated via the Kubernete Service. Therefore, the
	// cert is generated to certify the endpoint(s) defined by the "service".
```
We need to make clearly explain exactly what the Service is being used for.

If I'm not wrong then all we're expecting to do with the passed in Service object is to use it's name and namespace to set the certificate's SAN([DNSName](https://github.com/golang/go/blob/go1.10/src/crypto/x509/x509.go#L710)) to the Service FQDN `name.namespace.svc.cluster.local`
```Go
certTmpl := x509.Certificate{
...
		DNSNames:     []string{fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace)},
...
	}
```

So change this to something like:
```Go
// The passed in "service" is used to set the Subject Alternative Names(SAN) for the certificate. 
// We assume that the deployed applications are typically communicated with via a Kubernetes Service.
// The SAN is set to the FQDN of the service `<service-name>.<service-namespace>.svc.cluster.local`.
```

Reply:
yeah, i should make it a bit more explicit.